src/lsp/cobol_lsp/lsp_completion.ml

```diff
@@ -48,22 +50,89 @@ let change ~(case: actual_case) = match case with
   | Uppercase -> String.uppercase
   | Lowercase -> String.lowercase
 
-let to_string qualnames =
-  List.flatten @@ List.rev_map (function
-      | Cobol_ptree.Name name -> [~&name]
-      | Qual (name,_) as qualname ->
-        [~&name; Pretty.to_string "%a" Cobol_ptree.pp_qualname qualname]
-      | _ -> []) qualnames
+let to_string = function
+  | Cobol_ptree.Name name -> [~&name]
+  | Qual (name,_) as qualname ->
+    [~&name; Pretty.to_string "%a" Cobol_ptree.pp_qualname qualname]
+  | _ -> []
 
-let qualnames_proposal ~filename pos group =
-  match Lsp_lookup.cobol_unit_at_position ~filename pos group with
+let approx_type_of_pic ({ category; _ }: picture) =
+  match category with
+  | National _ | Alphabetic _ | Alphanumeric _ -> Alphanum
+  | FixedNum { editions; _ } when (editions.basics <> [] ||
+                                   editions.floating <> None ||
+                                   editions.zerorepl <> None)
+    -> NumericEdited
+  | FloatNum { editions; _ } when editions <> []
+    -> NumericEdited
+  | FloatNum _ | FixedNum _ -> Numeric
+  | Boolean _ -> Any
+
+let approx_type_of_usage : usage -> approx_typing_info = function
+  | Binary _
+  | Binary_C_long _
+  | Binary_char _
+  | Binary_double _
+  | Binary_long _
+  | Binary_short _
+  | Float_long
+  | Float_short
+  | Float_binary _
+  | Float_decimal _
+  | Float_extended
+  | Packed_decimal _ -> Numeric
+  | Procedure_pointer
+  | Function_pointer _
+  | Pointer _
+  | Index
+  | Program_pointer _ -> Numeric
+  | National _ -> Alphanum
+  | Object_reference _
+  | Bit _ -> Any
+  | Display pic -> approx_type_of_pic pic
+
+let approx_type_of_datadef : data_definition -> (approx_typing_info * bool) = fun d ->
+  match d with
+  | Data_field { def; _ } -> begin
+      match ~&def.field_layout with
+      | Elementary_field { usage; _ } -> (approx_type_of_usage usage, false)
+      | Struct_field _ -> (Alphanum, true)
+    end
+  | Data_renaming { def; _} -> begin
+      match ~&def.renaming_layout with
+      | Renamed_elementary { usage } -> (approx_type_of_usage usage, false)
+      | Renamed_struct _ -> (Alphanum, true)
+    end
```

Comment:
`Data_field` and `Data_renaming` can probably be merged by further destructing the `def`s.